File: internal/store/token_bucket.go
package store

import (
	"context"
	_ "embed"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/sohipan21/distributed-rate-limiter/internal/limiter"
)

//go:embed lua/token_bucket.lua
var tokenBucketScript string

// TokenBucket is the real redis-backed token bucket: check-and-update runs
// as one atomic lua script inside redis, so concurrent requests can't lose
// updates the way NaiveTokenBucket does
type TokenBucket struct {
	rdb    *redis.Client
	cfg    limiter.Config
	script *redis.Script
	opts   options
}

func NewTokenBucket(rdb *redis.Client, cfg limiter.Config, opts ...Option) *TokenBucket {
	if err := cfg.Validate(); err != nil {
		panic(err)
	}
	tb := &TokenBucket{rdb: rdb, cfg: cfg, script: redis.NewScript(tokenBucketScript)}
	for _, fn := range opts {
		fn(&tb.opts)
	}
	return tb
}

func (t *TokenBucket) Allow(key string) limiter.Decision {
	if !t.opts.breaker.allow() {
		return degradedDecision(t.opts.mode)
	}
	burst := t.cfg.Burst
	if burst <= 0 {
		burst = t.cfg.Limit
	}
	start := time.Now()
	res, err := t.script.Run(context.Background(), t.rdb,
		[]string{key},
		t.cfg.Limit, t.cfg.Window.Microseconds(), burst,
	).Int64Slice()
	if t.opts.observer != nil {
		t.opts.observer.ObserveRedis(string(limiter.TokenBucketAlgorithm), time.Since(start), err == nil)
	}
	if err != nil {
		t.opts.breaker.failure()
		return degradedDecision(t.opts.mode)
	}
	t.opts.breaker.success()
	return decode(res)
}

// both scripts return {allowed, remaining, retry_after_us, reset_at_us};
// reset_at is an absolute redis-clock timestamp, so every node agrees on it
func decode(res []int64) limiter.Decision {
	if len(res) != 4 {
		return limiter.Decision{}
	}
	return limiter.Decision{
		Allowed:    res[0] == 1,
		Remaining:  int(res[1]),
		RetryAfter: time.Duration(res[2]) * time.Microsecond,
		ResetAt:    time.UnixMicro(res[3]),
	}
}

File: internal/store/lua/token_bucket.lua
-- atomic token bucket. redis runs the whole script as one unit, so the
-- read-compute-write race in the naive store can't happen here.
-- KEYS[1] = bucket key
-- ARGV = limit, window (microseconds), burst
-- returns {allowed, remaining, retry_after_us, reset_at_us}

-- redis TIME is the single clock; node clocks never enter the math.
-- calling TIME before writes is fine on redis >= 5 (effect replication).
-- all time math in integer microseconds: fits a double exactly, and
-- avoids lua's %.14g tostring mangling large timestamps
local time = redis.call('TIME')
local now = time[1] * 1000000 + time[2]

local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local burst = tonumber(ARGV[3])
local rate = limit / window -- tokens per microsecond

local state = redis.call('HMGET', KEYS[1], 'tokens', 'last')
local tokens = burst
if state[1] then
  tokens = tonumber(state[1])
  local elapsed = now - tonumber(state[2])
  if elapsed > 0 then
    tokens = math.min(burst, tokens + elapsed * rate)
  end
end

local allowed = 0
local retry_after = 0
if tokens >= 1 then
  tokens = tokens - 1
  allowed = 1
else
  retry_after = math.ceil((1 - tokens) / rate)
end
local reset_at = now + math.ceil((burst - tokens) / rate)

-- format explicitly: letting the bridge stringify numbers goes through
-- %.14g, which corrupts microsecond timestamps
redis.call('HSET', KEYS[1],
  'tokens', string.format('%.17g', tokens),
  'last', string.format('%d', now))
-- state is reconstructible once the bucket refills, so let idle keys die
redis.call('PEXPIRE', KEYS[1], math.ceil(burst / rate / 1000))

return {allowed, math.floor(tokens), retry_after, reset_at}
